src/AssemblerReads.cpp

```diff
@@ -183,14 +184,90 @@ void Assembler::writeReadsSummary()
-        csv << reads.getMetaData(readId, "ch") << ",";
-        csv << reads.getMetaData(readId, "start_time") << ",";
+        csv << reads->getMetaData(readId, "runid") << ",";
+        csv << reads->getMetaData(readId, "sampleid") << ",";
+        csv << reads->getMetaData(readId, "read") << ",";
+        csv << reads->getMetaData(readId, "ch") << ",";
+        csv << reads->getMetaData(readId, "start_time") << ",";
 
         // End of the line for this read.
         csv << "\n";
-     }
+    }
+}
+
+
+void Assembler::adjustCoverage(uint64_t minReadLength, uint64_t desiredCoverage) {
+    if (desiredCoverage == 0) {
+        // Not specified. Skip.
+        return;
+    }
+
+    // Need to compute histograms to adjust coverage.
+    reads->computeReadLengthHistogram();
+
+    cout << timestamp << "Adjusting for desired coverage." << endl;
+    cout << "Desired Coverage: " << desiredCoverage << endl;
+    uint64_t cumulativeBaseCount = reads->getTotalBaseCount();
+
+    if (desiredCoverage > cumulativeBaseCount) {
+        throw runtime_error(
+            "With a Reads.minReadLength of " + to_string(minReadLength) + ","
+            "the total available coverage (" + to_string(cumulativeBaseCount) +
+            ") is lesser than the desired coverage (" + to_string(desiredCoverage) +
+            "). Try reducing Reads.minReadLength if appropriate or get more coverage."
+        );
+    }
+
+    uint64_t newMinReadLength = std::numeric_limits<uint64_t>::max();
+
+    const auto& binnedHistogram = reads->getBinnedHistogram();
+    for (uint64_t bin = 0; bin < binnedHistogram.size(); bin++) {
+        const auto& histogramBin = binnedHistogram[bin];
+        const uint64_t baseCount = histogramBin.second;
+
+        if (cumulativeBaseCount > desiredCoverage) {
+            cumulativeBaseCount -= baseCount;
+            continue;
+        }
+
+        newMinReadLength = max(uint64_t(0), bin - 1) * 1000;
+        break;
+    }
+
+    if (newMinReadLength == std::numeric_limits<uint64_t>::max()) {
+        throw runtime_error(
+            "Reads.desiredCoverage is set to " + to_string(desiredCoverage) +
+            " which is very low. Please check the value again."
+        );
+    }
+
+    SHASTA_ASSERT(newMinReadLength >= minReadLength);
+
+    cout << "Setting minReadLength to " + to_string(newMinReadLength) + 
+        " to get desired coverage." << endl;
+
+    // Rename existing memory mapped files to avoid overwriting data.
+    reads->renameWithSuffix("-OLD");
+
+    unique_ptr<Reads> newReads = make_unique<Reads>();
+    newReads->createNew(
+        largeDataName("Reads"),
+        largeDataName("ReadNames"),
+        largeDataName("ReadMetaData"),
+        largeDataName("ReadRepeatCounts"),
+        largeDataName("ReadFlags"),
+        largeDataPageSize
+    );
+
+    newReads->copyDataForReadsLongerThan(
+        getReads(),
+        newMinReadLength
+    );
+
+    reads->remove();
+    
+    reads.swap(newReads);
```

Comment:
All we need is `reads = newReads`, right? The call to `swap` also assigns a new value to `newReads`, but that is superfluous as `newReads` goes immediately out of scope. So the call to `swap` seems confusing to me.

Another alternative would be do to without the `unique_ptr` altogether, and have `class Reads` provide a `swap` member function similar in spirit to `std::vector::swap()`, which would just swap the pointers between the `*this` and the `Reads` passed as argument.

Reply:
In the latter alternative, all changes caused by `reads` becoming a `unique_ptr` would no longer be needed.

Reply:
Yeah. I'll simply do `reads = std::move(newReads);` instead.